cellmlmanip/model.py

```diff
@@ -436,18 +470,37 @@ def __connect(self, connection):
         """
         cmp_name_1, var_att_1, cmp_name_2, var_att_2 = self.__get_connection_endpoints(connection)
 
-        # Determine which are the source and target variables, and connect from->to
-        if (('public_interface', 'out') in var_att_1.items() or
-                ('private_interface', 'out') in var_att_1.items()) and (
-                    ('public_interface', 'in') in var_att_2.items() or
-                    ('private_interface', 'in') in var_att_2.items()):
-            return self.__connect_with_direction(cmp_name_1, var_att_1, cmp_name_2, var_att_2)
-        elif (('public_interface', 'out') in var_att_2.items() or
-              ('private_interface', 'out') in var_att_2.items()) and (
-                  ('public_interface', 'in') in var_att_1.items() or
-                  ('private_interface', 'in') in var_att_1.items()):
-            return self.__connect_with_direction(cmp_name_2, var_att_2, cmp_name_1, var_att_1)
-        raise RuntimeError("Cannot determine the source & target for connection %s" % connection)
+        # keys for lookup
+        pub = 'public_interface'
+        pri = 'private_interface'
+
+        def __are_siblings(name_a, name_b):
+            return self.components[name_a].parent == self.components[name_b].parent
+
+        def __parent_of(parent_name, child_name):
+            return parent_name == self.components[child_name].parent
+
+        def __has_interface(variable_a, key_a, value_a, variable_b, key_b, value_b):
+            return (key_a, value_a) in variable_a.items() and (key_b, value_b) in variable_b.items()
+
+        # if one component is encapsulated by the other
```

Comment:
The relevant lines from the spec are

> The set of all components immediately encapsulated by the current component is the encapsulated set.

> Other components encapsulated by the same parent make up the sibling set.

> The interface exposed to the parent component and components in the sibling set is defined by the public_interface attribute. The private_interface attribute defines the interface exposed to components in the encapsulated set. Each interface has three possible values: "in", "out", and "none", where "none" indicates the absence of an interface.

might be good to quote those in a comment!